### PR TITLE
Define Halo confidence event contract

### LINKED ISSUE #123: Define Halo confidence event contract
## What this is about
Halo now has real RGB ingest, captured datasets, and a measurable recognition baseline. The next step is a typed confidence event contract so backend detection output, evidence logging, and the GCS demo all speak the same shape.

## Scope
- define the Halo detection/confidence event shape with source id/name, timestamps, label/type, confidence, confidence level, frame id, frame index, optional bounding region or location hint, and optional evidence reference
- provide a backend conversion path from RGB recognition candidates/evaluation output into the event shape
- provide viewer-side parsing/normalization utilities for the same contract
- cover defaults, invalid payloads, confidence bounds, optional geometry/evidence fields, and round-trip compatibility in tests

## Done means
- perception and viewer code agree on one event contract
- malformed or out-of-range payloads fail or normalize predictably
- the contract can feed the existing GCS detection/confidence display patterns without requiring a full mission-state implementation
- evidence logging and the GCS demo can build on this without redefining the detection payload

## Out of scope
- evidence persistence
- GCS visual redesign
- trained ML model work
- thermal/audio fusion changes
- drone or rover hardware integration

### PR DESCRIPTION
**Pull request by SAKETH11111**
## Summary
- add a canonical Halo confidence event contract and candidate conversion helper in perception
- add viewer normalization for Halo confidence events, including optional region, location, evidence, and recognition metadata handling
- add backend and viewer tests covering defaults, invalid payloads, confidence bounds, and contract compatibility

## Validation
- `PYTHONPATH=software/edge/src/aeris_perception:software/edge/src/aeris_msgs python3 -m pytest software/edge/src/aeris_perception/test/test_halo_confidence_event.py software/edge/src/aeris_msgs/test -q`
- `cd software/viewer && node --test src/lib/ros/*.test.mjs src/lib/*.test.mjs`
- `python3 -m compileall software/edge/src/aeris_perception/aeris_perception`

## Notes
- the broader perception pytest command currently stops during collection in this environment because `numpy` is not installed locally: `PYTHONPATH=software/edge/src/aeris_perception:software/edge/src/aeris_msgs python3 -m pytest software/edge/src/aeris_perception/test software/edge/src/aeris_msgs/test -q`

Closes #123

## Summary by CodeRabbit

* **New Features**
  * Added Halo confidence event framework for RGB recognition outputs with confidence tiering and structured detection metadata
  * Implemented event normalization and validation pipelines for improved data consistency across components

* **Tests**
  * Added comprehensive test coverage for event parsing, serialization, and validation

[Review Change Stack](https://app.coderabbit.ai/change-stack/Aeris-Drones/aeris/pull/124?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

<h3>Greptile Summary</h3>

This PR introduces the canonical Halo confidence event contract: a Python dataclass with parser/serializer/candidate-converter in `aeris_perception`, a matching JavaScript normalizer with TypeScript declarations in the viewer, and test suites for both. Several previously flagged issues (bool coercion, ns→ms timestamp, null `requireFiniteNumber`, field-name in `_optional_string` errors) have been addressed.

- **Python backend** (`halo_confidence_event.py`): well-guarded validators (bool rejection, finite-float, string-from-int), clean round-trip via `to_dict`/`parse_halo_confidence_event`, and a candidate-conversion helper.
- **Viewer normalizer** (`haloConfidenceEvents.js`): BigInt precision-safe timestamp handling, confidence-level tiering, and display-timestamp derivation with monotonic-clock fallback.
- **Remaining asymmetry**: the three optional recognition sub-fields (`latency_ms`, `source_timestamp_ns`, `confidence_components`) use `!== undefined` guards in `parseRecognition`, so a `null` value throws in JS but is silently accepted in Python — a payload valid on the backend is rejected by the viewer.

<h3>Confidence Score: 4/5</h3>

Safe to merge after fixing the null-guard asymmetry in parseRecognition; all other logic is consistent between Python and JavaScript.

The viewer's parseRecognition uses `!== undefined` for three optional fields, so a null value (valid and silently ignored on the backend) causes the viewer to throw. A backend payload with, e.g., `latency_ms: null` would break the viewer's normalizer, yet pass Python validation without error.

software/viewer/src/lib/ros/haloConfidenceEvents.js — the `parseRecognition` function's optional-field guards need `!= null` instead of `!== undefined`.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| software/edge/src/aeris_perception/aeris_perception/halo_confidence_event.py | New canonical Halo confidence event contract with dataclass, parser, serializer, and candidate converter. Validators are thorough (bool guard, finite-float, int-from-string), round-trip and optional-field handling look correct. |
| software/viewer/src/lib/ros/haloConfidenceEvents.js | New viewer normalizer for Halo confidence events. Previous issues (bool coercion, ns→ms conversion, requireFiniteNumber null) addressed, but three optional recognition sub-field guards use `!== undefined` instead of `!= null`, causing JS to throw on null values that Python silently accepts. |
| software/viewer/src/lib/ros/haloConfidenceEvents.d.ts | TypeScript declaration file for the viewer normalizer. Types accurately reflect optional fields; `timestampNs` correctly typed as `string` to avoid precision loss. |
| software/edge/src/aeris_perception/test/test_halo_confidence_event.py | Comprehensive pytest coverage: candidate conversion, round-trip, clamping, and a parametrized invalid-payload suite. All previously discussed edge cases (bool confidence, float frame_index, non-string source_uri) are covered. |
| software/viewer/src/lib/ros/haloConfidenceEvents.test.mjs | Node test-runner tests for canonical mapping, defaults, string location hints, label-only derivation, and invalid payload rejection. Covers boolean frame_index/timestamp_ns and null confidence cases added in this PR. |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant C as RecognitionCandidate
    participant P as halo_confidence_event.py
    participant M as ROS Message (JSON)
    participant J as haloConfidenceEvents.js
    participant V as Viewer UI

    C->>P: halo_confidence_event_from_candidate(candidate)
    P->>P: "validate fields (_require_*, _clamp01)"
    P->>P: build HaloConfidenceEvent dataclass
    P->>M: serialize_halo_confidence_event() → dict
    Note over M: timestamp_ns as string,<br/>source_timestamp_ns as string

    M->>J: normalizeHaloConfidenceEventMessage(rawMessage)
    J->>J: requireString / requireBigInt / requireFiniteNumber
    J->>J: deriveDisplayTimestampMs (ns→ms, monotonic fallback)
    J->>J: parseRegion / parseLocationHint / parseRecognition
    J->>V: NormalizedHaloConfidenceDetection
```

<details><summary><h3>Comments Outside Diff (2)</h3></summary>

1. `software/viewer/src/lib/ros/haloConfidenceEvents.js`, line 752-758 ([link](https://github.com/aeris-drones/aeris/blob/daf8848eae26def5d905f50173098358eadbbb66/software/viewer/src/lib/ros/haloConfidenceEvents.js#L752-L758)) 

   <a href="#">P1</a> **`requireFiniteNumber` silently accepts `null`, producing `0` for required fields**

   `Number(null) === 0` which is finite, so `requireFiniteNumber(null, "confidence")` returns `0` without throwing. A message with `confidence: null` would then produce a normalized detection with `confidence = 0` — no error, wrong data. The Python counterpart correctly rejects `None` (`float(None)` raises `TypeError`, caught and re-raised as `ValueError`). The same silent-zero path also affects `recognition.latency_ms` and `recognition.source_timestamp_ns` via the `!== undefined` guards (since `null !== undefined` is `true`), where Python skips null optional fields entirely (`if latency_ms is not None:`). Adding a `rawValue === null` guard before the `Number()` call (or using `rawValue == null` to cover both) would align with Python's behavior.

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: software/viewer/src/lib/ros/haloConfidenceEvents.js
   Line: 752-758

   Comment:
   **`requireFiniteNumber` silently accepts `null`, producing `0` for required fields**

   `Number(null) === 0` which is finite, so `requireFiniteNumber(null, "confidence")` returns `0` without throwing. A message with `confidence: null` would then produce a normalized detection with `confidence = 0` — no error, wrong data. The Python counterpart correctly rejects `None` (`float(None)` raises `TypeError`, caught and re-raised as `ValueError`). The same silent-zero path also affects `recognition.latency_ms` and `recognition.source_timestamp_ns` via the `!== undefined` guards (since `null !== undefined` is `true`), where Python skips null optional fields entirely (`if latency_ms is not None:`). Adding a `rawValue === null` guard before the `Number()` call (or using `rawValue == null` to cover both) would align with Python's behavior.

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

2. `software/viewer/src/lib/ros/haloConfidenceEvents.js`, line 916-937 ([link](https://github.com/aeris-drones/aeris/blob/f21fb9cd305dbd47dac5eefbd98ee806be126959/software/viewer/src/lib/ros/haloConfidenceEvents.js#L916-L937)) 

   <a href="#">P1</a> **`null` optional recognition fields throw instead of being skipped**

   All three optional-field guards inside `parseRecognition` use `!== undefined`, so a `null` value for `latency_ms`, `source_timestamp_ns`, or `confidence_components` passes the guard and is then handed to `requireFiniteNumber`/`requireBigInt`/`requireObject`, which all throw for `null`. Python's `_normalize_recognition` uses `if latency_ms is not None:`, which silently skips `null`/`None` for all three fields. A backend payload containing `recognition: { …, latency_ms: null }` is therefore valid on the Python side and raises in the viewer, breaking the contract. Change each guard to `!= null` (or `!== undefined && !== null`) to match Python's behaviour.

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: software/viewer/src/lib/ros/haloConfidenceEvents.js
   Line: 916-937

   Comment:
   **`null` optional recognition fields throw instead of being skipped**

   All three optional-field guards inside `parseRecognition` use `!== undefined`, so a `null` value for `latency_ms`, `source_timestamp_ns`, or `confidence_components` passes the guard and is then handed to `requireFiniteNumber`/`requireBigInt`/`requireObject`, which all throw for `null`. Python's `_normalize_recognition` uses `if latency_ms is not None:`, which silently skips `null`/`None` for all three fields. A backend payload containing `recognition: { …, latency_ms: null }` is therefore valid on the Python side and raises in the viewer, breaking the contract. Change each guard to `!= null` (or `!== undefined && !== null`) to match Python's behaviour.

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
Fix the following 1 code review issue. Work through them one at a time, proposing concise fixes.

---

### Issue 1 of 1
software/viewer/src/lib/ros/haloConfidenceEvents.js:916-937
**`null` optional recognition fields throw instead of being skipped**

All three optional-field guards inside `parseRecognition` use `!== undefined`, so a `null` value for `latency_ms`, `source_timestamp_ns`, or `confidence_components` passes the guard and is then handed to `requireFiniteNumber`/`requireBigInt`/`requireObject`, which all throw for `null`. Python's `_normalize_recognition` uses `if latency_ms is not None:`, which silently skips `null`/`None` for all three fields. A backend payload containing `recognition: { …, latency_ms: null }` is therefore valid on the Python side and raises in the viewer, breaking the contract. Change each guard to `!= null` (or `!== undefined && !== null`) to match Python's behaviour.


`````

</details>

<sub>Reviews (3): Last reviewed commit: ["Preserve Halo event timestamp precision"](https://github.com/aeris-drones/aeris/commit/f21fb9cd305dbd47dac5eefbd98ee806be126959) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=34496848)</sub>

### CODE REVIEW

File: software/edge/src/aeris_perception/aeris_perception/halo_confidence_event.py [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+    return raw_value.strip()
+
+
+def _optional_string(raw_value: Any) -> str | None:
+    if raw_value is None:
+        return None
+    if not isinstance(raw_value, str):
+        raise ValueError("optional string field must be a string when provided")
+    stripped = raw_value.strip()
+    return stripped or None
```
[the remark this answers is not in the document]
Comment (SAKETH11111) [reply software/edge/src/aeris_perception/aeris_perception/halo_confidence_event.py]: Updated this. Optional string validation now includes the field name, and the parser covers a non-string source_uri case.

File: software/viewer/src/lib/ros/haloConfidenceEvents.js [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+  return normalized === "" ? undefined : normalized;
+}
+
+function requireInt(rawValue, fieldName, minimum = 0) {
+  const value = Number(rawValue);
+  if (!Number.isInteger(value) || value < minimum) {
+    throw new Error(`Invalid Halo confidence event: ${fieldName} must be an integer >= ${minimum}`);
+  }
+  return value;
+}
```
[the remark this answers is not in the document]
Comment (SAKETH11111) [reply software/viewer/src/lib/ros/haloConfidenceEvents.js]: Fixed this. The viewer integer validator now rejects boolean values before numeric coercion, matching the backend parser.

File: software/viewer/src/lib/ros/haloConfidenceEvents.js [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+      frameId,
+      frameIndex,
+      timestampNs,
+      detectionType: resolvedDetectionType,
+      region,
+    }),
+    sensorType: "rgb",
+    confidence,
+    confidenceLevel: normalizeConfidenceLevel(message.confidence_level, confidence),
+    timestamp: Math.floor(timestampNs / 1_000),
```
[the remark this answers is not in the document]
Comment (SAKETH11111) [reply software/viewer/src/lib/ros/haloConfidenceEvents.js]: Fixed this. The viewer display timestamp now converts nanoseconds to milliseconds while preserving timestampNs on the normalized detection.

File: software/viewer/src/lib/ros/haloConfidenceEvents.js [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+  return value;
+}
+
+function requireFiniteNumber(rawValue, fieldName) {
+  const value = Number(rawValue);
+  if (!Number.isFinite(value)) {
+    throw new Error(`Invalid Halo confidence event: ${fieldName} must be numeric`);
+  }
+  return value;
+}
```
[the remark this answers is not in the document]
Comment (SAKETH11111) [reply software/viewer/src/lib/ros/haloConfidenceEvents.js:86]: Handled this. Viewer numeric parsing now rejects null, undefined, and boolean values before coercion, and the invalid-payload tests cover the null confidence case.

File: software/edge/src/aeris_perception/aeris_perception/halo_confidence_event.py [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+
+
+def _require_finite_float(raw_value: Any, field_name: str) -> float:
+    try:
+        value = float(raw_value)
+    except (TypeError, ValueError) as exc:
+        raise ValueError(f"{field_name} must be numeric") from exc
+    if not math.isfinite(value):
+        raise ValueError(f"{field_name} must be numeric")
+    return value
```
[the remark this answers is not in the document]
Comment (SAKETH11111) [reply software/edge/src/aeris_perception/aeris_perception/halo_confidence_event.py:389]: Handled this. Backend integer parsing now only accepts real integers or integer strings, finite-float parsing rejects booleans, and tests cover fractional and boolean payloads.

**Comment by SAKETH11111** [comment]
Addressed the timestamp precision issue. Nanosecond fields are preserved as strings across the JSON/JS boundary, display timestamps stay in milliseconds, and BigInt is used only for timestamp math.
